src/hotspot/share/opto/matcher.cpp

```diff
@@ -2320,6 +2320,16 @@ bool Matcher::find_shared_visit(MStack& mstack, Node* n, uint opcode, bool& mem_
 }
 
 void Matcher::find_shared_post_visit(Node* n, uint opcode) {
+  if (n->is_Vector() && (n->req() == 4) && n->in(3)->is_VectorMask()) {
+    // Handle op with mask
+    Node* pair = new BinaryNode(n->in(1), n->in(2));
+    Node* mask = n->in(3);
+    n->set_req(1, pair);
+    n->set_req(2, mask);
+    n->del_req(3);
+    return;
+  }
+
```

Comment:
If we add a new flag for masked nodes, it may simplify the checking logic for masking operations.  e.g. 
node.hpp:
class Node {
  // Flags are sorted by usage frequency.
  enum NodeFlags {
    Flag_is_Copy                     = 1 << 0, // should be first bit to avoid shift
    Flag_rematerialize               = 1 << 1,
    Flag_needs_anti_dependence_check = 1 << 2,
    Flag_is_macro                    = 1 << 3,
    Flag_is_Con                      = 1 << 4,
    Flag_is_cisc_alternate           = 1 << 5,
    Flag_is_dead_loop_safe           = 1 << 6,
    Flag_may_be_short_branch         = 1 << 7,
    Flag_avoid_back_to_back_before   = 1 << 8,
    Flag_avoid_back_to_back_after    = 1 << 9,
    Flag_has_call                    = 1 << 10,
    Flag_is_reduction                = 1 << 11,
    Flag_is_scheduled                = 1 << 12,
    **Flag_has_vector_mask_set**         = 1 << 13,

Reply:
Good suggestion! Thanks! So do you mean using the new flag instead of `"n->in(3)->is_VectorMask()" ` ? I'm not very clarified about what the checking logic for masking operations you mean. Could you please elaborate more about this? Thanks so much!

Reply:
Yes, setting a new flag over IR node could simplify the checks , and could be used in matcher patterns also.

Reply:
OK, I will try to use it. Thanks for your advice!

Reply:
Hi @jatin-bhateja , I'v updated the patch according to your advice. Please help to look at it again. Thanks so much!